api/src/OpenTelemetry/Attributes/Map.hs

```diff
@@ -0,0 +1,60 @@
+{-# LANGUAGE BangPatterns #-}
+{-# LANGUAGE DefaultSignatures #-}
+{-# LANGUAGE DeriveAnyClass #-}
+{-# LANGUAGE DeriveDataTypeable #-}
+{-# LANGUAGE DeriveGeneric #-}
+{-# LANGUAGE DerivingStrategies #-}
+{-# LANGUAGE ExistentialQuantification #-}
+{-# LANGUAGE GeneralizedNewtypeDeriving #-}
+{-# LANGUAGE LambdaCase #-}
+{-# LANGUAGE StrictData #-}
+{-# LANGUAGE TypeFamilies #-}
+
+{- |
+Module      :  OpenTelemetry.Attributes.Attribute
+Copyright   :  (c) Ian Duncan, 2021
+License     :  BSD-3
+Description :  Key-value pair metadata used in 'OpenTelemetry.Trace.Span's, 'OpenTelemetry.Trace.Link's, and 'OpenTelemetry.Trace.Event's
+Maintainer  :  Ian Duncan
+Stability   :  experimental
+Portability :  non-portable (GHC extensions)
+-}
+module OpenTelemetry.Attributes.Map (
+  AttributeMap,
+  insertByKey,
+  insertAttributeByKey,
+  lookupByKey,
+  lookupAttributeByKey,
+  module H,
+) where
+
+import Data.HashMap.Strict as H
+import Data.Text (Text)
+import OpenTelemetry.Attributes.Attribute (
+  Attribute,
+  FromAttribute (fromAttribute),
+  ToAttribute (toAttribute),
+ )
+import OpenTelemetry.Attributes.Key (
+  Key (Key),
+ )
+import Prelude hiding (lookup, map)
+
+
+type AttributeMap = H.HashMap Text Attribute
+
+
+insertByKey :: ToAttribute a => Key a -> a -> AttributeMap -> AttributeMap
```

Comment:
if we had called it `AttributeName`, then this would be `insertByName`, which I think sounds nice?

Reply:
Another option: remove `ByKey` suffix and just call it `insert`.

Reply:
Actually I want to rename this to `insert` and to remove the original `insert`. I kept `insert` because of compatibility.